### PR TITLE
VReplication: improve reliability of log management

### PR DESCRIPTION
**Pull request by mattlord**
## Description

In helping someone debug a VReplication failure, we had seen the following:
1. We failed to create a `_vt.vreplication_log` record and the apparent cause was that the message was too long
2. The truncated error message seen in the `_vt.vreplication` record showed the beginning of a `vcopier` bulk `INSERT` query that failed — truncated at 1000 bytes — but not the *reason why / error it generated*
3. This `_vt.vreplication_log` record generation failure may have caused problems in the workflow (e.g. if this happens after the point of no return in a `SwitchTraffic` then we'll be left in an inconsistent state that needs manual cleanup)

We address all of these issues with the following changes in this PR:
1. We truncate the message as needed before inserting it into the `_vt.vreplication_log` table
2. That truncation happens in the middle of the message so that we don't lose the error shown e.g. after a failed query
3. We no longer consider failing to write a `_vt.vreplication_log` record as "fatal" — meaning that the workflow operation does not stop and return the log generation error — and instead only logs an error in the `vttablet` log

I also moved the existing truncation for the `_vt.vreplication.message` field over to this new truncation method so that we have the error message there as well.

## Related Issue(s)

**I didn't create an issue since this felt more like a minor internal cleanup, but I can create one quickly if others disagree.**

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vttablet/tabletmanager/vreplication/utils.go [hunk 1]
```diff
@@ -82,46 +91,53 @@ func getLastLog(dbClient *vdbClient, vreplID int32) (id int64, typ, state, messa
...
 		// handles case where current state is Running, controller restarts after an error and initializes the state Running
-		return nil
+		return
 	}
 	var query string
 	if id > 0 && message == lastLogMessage {
 		query = fmt.Sprintf("update %s.vreplication_log set count = count + 1 where id = %d", sidecar.GetIdentifier(), id)
 	} else {
 		buf := sqlparser.NewTrackedBuffer(nil)
+		// We perform the truncation, if needed, in the middle of the message as the end of the message is likely to
```
Comment (rohit-nayak-ps) [inline go/vt/vttablet/tabletmanager/vreplication/utils.go:111]: You should just extract this logic into a  helper, which just does the fitting to size. That will also simplify the unit test ...
Also the for loop probably can be replaced to make it O(1). 

```
const prefixFactor = 0.4

func FitToSize(message string, maxLen int, truncationIndicator string) string {
	l := len(message)
	availableLen := maxLen - len(truncationIndicator)

	if l <= availableLen {
		return message
	}

	prefixLen := int(float64(availableLen) * prefixFactor)
	suffixLen := availableLen - prefixLen

	return message[:prefixLen] + truncationIndicator + message[l-suffixLen:]
}
```

File: go/vt/vttablet/tabletmanager/vreplication/utils.go [hunk 2]
```diff
@@ -33,8 +33,17 @@ import (
...
+	// This comes from the fact that the message column in the vreplication_log table is of type TEXT.
+	// See: go/vt/sidecardb/schema/vreplication/vreplication_log.sql
+	// https://dev.mysql.com/doc/refman/en/string-type-syntax.html and
+	// https://dev.mysql.com/doc/refman/en/storage-requirements.html#data-types-storage-reqs-strings
+	maxVReplicationLogMessageLen = 65535
 )
 
+// vrepliationLogTruncationStr is the string that is used to indicate that a message has been
+// truncated, in the middle, before being inserted into the vreplication_log table.
+var vrepliationLogTruncationStr = fmt.Sprintf(" ... %s ... ", sqlparser.TruncationText)
```
Comment (rohit-nayak-ps) [inline go/vt/vttablet/tabletmanager/vreplication/utils.go:45]: typo => vreplication ...

File: go/vt/vttablet/tabletmanager/vreplication/utils.go [hunk 3]
```diff
@@ -33,8 +33,17 @@ import (
 
 const (
 	vreplicationLogTableName = "vreplication_log"
+	// This comes from the fact that the message column in the vreplication_log table is of type TEXT.
```
Comment (rohit-nayak-ps) [inline go/vt/vttablet/tabletmanager/vreplication/utils.go:36]: Do we need this verbosity in the comments?  We can just say that this is to ensure message fits in the column, no?

**Review by rohit-nayak-ps — COMMENTED** [review]
Looks good! 

Had a deja vu about this: I thought we had added this truncation a long time ago, because we had seen this error before for long queries. 

Turns out we had done it for `_vt.vreplication` while setting the ubiquitous `message` column:

```
func (ct *controller) setMessage(dbClient binlogplayer.DBClient, message string) error {
	ct.blpStats.History.Add(&binlogplayer.StatsHistoryRecord{
		Time:    time.Now(),
		Message: message,
	})
	query := fmt.Sprintf("update _vt.vreplication set message=%v where id=%v", encodeString(binlogplayer.MessageTruncate(message)), ct.id)
	if _, err := dbClient.ExecuteFetch(query, 1); err != nil {
		return fmt.Errorf("could not set message: %v: %v", query, err)
	}
	return nil
}
```

Maybe we can just update this function to use your idea of logging enough of the prefix and suffix and use it in this PR as well.

File: go/vt/binlog/binlogplayer/dbclient.go [hunk 4]
```diff
@@ -129,10 +130,14 @@ func LogError(msg string, err error) {
 
 // LimitString truncates string to specified size
 func LimitString(s string, limit int) string {
-	if len(s) > limit {
+	ls, err := textutil.TruncateText(s, limit, textutil.TruncationLocationMiddle, "...")
+	if err != nil { // Fallback to simple truncation
+		if len(s) <= limit {
```
Comment (ajm188) [inline go/vt/binlog/binlogplayer/dbclient.go:135]: should we do this check before calling `TruncateText`? we'd potentially save some computation

**Review by ajm188 — APPROVED** [review]

File: go/vt/binlog/binlogplayer/dbclient.go (hunk 4 above)
Comment (mattlord) [reply go/vt/binlog/binlogplayer/dbclient.go:135]: The util function does that, which is why I didn't. But I could do it first here too....

Comment (ajm188) [reply go/vt/binlog/binlogplayer/dbclient.go:135]: Ah that works for me! Didn't know that

**Review by rohit-nayak-ps — APPROVED** [review]
Nice work.

**Pull request merged**